### PR TITLE
eth: add debug_storageRangeAt

### PR DESCRIPTION
**Pull request by fjl**
The new RPC method allows downloading storage in chunks.

`debug_storageRangeAt` takes parameters `blockHash`, `txIndex`, `account`, `startKey`, `limit`. The `startKey` applies to the hashed key (instead of the preimage as originally suggested in #3407) . The returned storage values contain the preimage if available. The returned object also contains the `nextKey` field, which will be non-null if there are more keys after the range that was returned.

Example:

```js
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "debug_storageRangeAt",
  "params": [
    "0xced82545131cdb6253f183a62dc3eaf81259be5cc744767798bcfb95664dc32c", // blockHash
    4,                                                                    // txIndex
    "0xB4A5BB81C6962Fa87Af6886ABE59b3cc56ec44b8",                         // account
    "0x",                                                                 // startKey
    2                                                                     // limit
  ]
}
// Response:
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "storage": {
      "0x1ab0c6948a275349ae45a06aad66a8bd65ac18074615d53676c09b67809099e0": {
        "key": "0x405787fa12a823e0f2b7631cc41b3ba8828b3321ca811111fa75cd3aa3bb5ace",
        "value": "0x00000000000000000000009413739a6ef2d14b5433465dbeffafbc2f2d610551"
      },
      "0x1d3954ea05fac9c4f44f290d32ed11be13922e2c8162b3594bb3bb0405546e8e": {
        "key": "0x2a359015afbfbe3b36db751c88f03aed63209927e8cf9ca7f0b69158e29e28f5",
        "value": "0x0000000000000000000000000000000000000000000000000000000000000004"
      }
    },
    "nextKey": "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563"
  }
}

// Continuing the download using nextKey as startKey:
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "debug_storageRangeAt",
  "params": [
    "0xced82545131cdb6253f183a62dc3eaf81259be5cc744767798bcfb95664dc32c",
    4,
    "0xB4A5BB81C6962Fa87Af6886ABE59b3cc56ec44b8",
    "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563",
    100
  ]
}
// The response has nextKey == null because there aren't any more values:
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "storage": {
      "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563": {
        "key": "0x0000000000000000000000000000000000000000000000000000000000000000",
        "value": "0x0000000000000000000000945e237bf1540eea146f5e830593d1d54c279616a9"
      },
      ... // 16 more entries elided
    },
    "nextKey": null
  }
}
```

We agreed on a preliminary spec in #3407. This supersedes that PR.
The earlier commits leading up to the new API add support for seeking in the trie iterator.

### CODE REVIEW

File: trie/encoding.go [hunk 1]
```diff
@@ -16,49 +16,54 @@
...
+//
+// COMPACT encoding is defined by the Ethereum Yellow Paper (it's called "hex prefix
+// encoding" there) and contains the bytes of the key and a flag. The high nibble of the
+// first byte contains the flag; the lowest bit encoding the oddness of the length and
+// the second-lowest encoding whether the node at the key is a value node. The low nibble
+// of the first byte is zero in the case of an even number of nibbles and the first nibble
+// in the case of an odd number. All remaining nibbles (now an even number) fit properly
+// into the remaining bytes. Compact encoding is used for nodes stored on disk.
+
+func hexToCompact(hex []byte) []byte {
```
Comment (Arachnid) [inline trie/encoding.go]: Could you define typedefs for `hex` and `compact` encoding?

File: trie/iterator.go [hunk 2]
```diff
@@ -85,25 +82,25 @@ type nodeIteratorState struct {
...
-	err error // Failure set in case of an internal error in the iterator
-
-	path []byte // Path to the current node
+	err   error                // Failure set in case of an internal error in the iterator
+	path  []byte               // Path to the current node
 }
 
-// NewNodeIterator creates an post-order trie iterator.
-func NewNodeIterator(trie *Trie) NodeIterator {
+// newNodeIterator creates an post-order trie iterator.
```
Comment (Arachnid) [inline trie/iterator.go]: preorder. :)

File: trie/iterator.go [hunk 3]
```diff
@@ -85,25 +82,25 @@ type nodeIteratorState struct {
...
+func newNodeIterator(trie *Trie, start []byte) NodeIterator {
```
Comment (Arachnid) [inline trie/iterator.go]: What was the motivation behind making this internal?

File: trie/iterator.go [hunk 4]
```diff
@@ -161,111 +161,132 @@ func (it *nodeIterator) Error() error {
...
 		ancestor := parent.hash
 		if (ancestor == common.Hash{}) {
 			ancestor = parent.parent
 		}
 		if node, ok := parent.node.(*fullNode); ok {
-			// Full node, iterate over children
-			for parent.child++; parent.child < len(node.Children); parent.child++ {
-				child := node.Children[parent.child]
+			// Full node, move to the first non-nil child.
+			i := parent.index + 1
```
Comment (Arachnid) [inline trie/iterator.go]: Why isn't this part of the loop initializer any longer?

File: trie/iterator_test.go [hunk 5]
```diff
@@ -140,6 +144,41 @@ var testdata2 = []struct{ k, v string }{
...
+	if err := checkIteratorOrder(testdata1[4:], it); err != nil {
+		t.Fatal(err)
+	}
+
+	// Seek to a non-existent key.
+	it = NewIterator(trie.NodeIterator([]byte("barc")))
+	if err := checkIteratorOrder(testdata1[1:], it); err != nil {
+		t.Fatal(err)
+	}
+}
```
Comment (Arachnid) [inline trie/iterator_test.go]: Can you add tests  for seeking to the empty string, and for seeking to a key past the end?

File: trie/encoding.go (hunk 1 above)
Comment (fjl) [inline trie/encoding.go]: No, because both types of key can be stored in `shortNode.Key`.

File: trie/iterator.go (hunk 3 above)
Comment (fjl) [inline trie/iterator.go]: The motivation was to have a single constructor for each iterator. Please see the commit message.

File: trie/iterator.go (hunk 4 above)
Comment (fjl) [inline trie/iterator.go]: Accident

File: trie/encoding.go (hunk 1 above)
Comment (Arachnid) [inline trie/encoding.go]: We could always use casts there. Typedefs would be helpful to avoid confusion elsewhere - though the function naming accomplishes some of that at least.

File: trie/iterator_test.go (hunk 5 above)
Comment (fjl) [inline trie/iterator_test.go]: No need to test empty string, all other tests use `NodeIterator(nil)` ;). I've added one that seeks to the end.

File: trie/encoding.go (hunk 1 above)
Comment (fjl) [inline trie/encoding.go]: I don't want to change it in this PR. We can do it in another PR.

File: eth/api_test.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+var dumper = spew.ConfigState{Indent: "    "}
+
+func TestStorageRangeAt(t *testing.T) {
+	// Create a state where account 0x010000... has a few storage entries.
+	var (
+		db, _    = ethdb.NewMemDatabase()
+		state, _ = state.New(common.Hash{}, db)
+		addr     = common.Address{0x01}
+		keys     = []common.Hash{
+			common.HexToHash("340dd630ad21bf010b4e676dbfa9ba9a02175262d1fa356232cfde6cb5b47ef2"),
```
Comment (Arachnid) [inline eth/api_test.go]: I assume these are the hashes of the keys in the array below? Could you add a note to that effect for anyone coming across this later?

File: trie/iterator.go [hunk 7]
```diff
@@ -155,111 +160,131 @@ func (it *nodeIterator) Error() error {
...
-
 	if !descend {
 		// If we're skipping children, pop the current node first
-		it.path = it.path[:it.stack[len(it.stack)-1].pathlen]
-		it.stack = it.stack[:len(it.stack)-1]
+		it.pop()
 	}
 
 	// Continue iteration to the next child
-outer:
```
Comment (karalabe) [inline trie/iterator.go]: :heart: I hate these labels so much :P

File: trie/iterator.go [hunk 8]
```diff
@@ -155,111 +160,131 @@ func (it *nodeIterator) Error() error {
...
 			}
 		} else if hash, ok := parent.node.(hashNode); ok {
 			// Hash node, resolve the hash child from the database
-			if parent.child < 0 {
-				parent.child++
+			if parent.index < 0 {
 				node, err := it.trie.resolveHash(hash, nil, nil)
 				if err != nil {
-					return err
+					return it.stack[len(it.stack)-1], &parent.index, it.path, err
```
Comment (karalabe) [inline trie/iterator.go]: Why return the previous "item" if peeking at the next one fails?

File: trie/encoding.go [hunk 9]
```diff
@@ -16,49 +16,54 @@
 
 package trie
 
-func compactEncode(hexSlice []byte) []byte {
+// Trie keys are dealt with in three distinct encodings:
+//
+// KEYBYTES encoding contains the actual key and nothing else. This encoding is the
+// input to most API functions.
+//
+// HEX encoding contains on byte for each nibble of the key and an optional trailing
```
Comment (karalabe) [inline trie/encoding.go]: onE byte

File: trie/encoding.go [hunk 10]
```diff
@@ -16,49 +16,54 @@
...
+// contains a value. Hex key encoding is used for nodes loaded in memory because it's
+// convenient to access.
+//
+// COMPACT encoding is defined by the Ethereum Yellow Paper (it's called "hex prefix
+// encoding" there) and contains the bytes of the key and a flag. The high nibble of the
+// first byte contains the flag; the lowest bit encoding the oddness of the length and
+// the second-lowest encoding whether the node at the key is a value node. The low nibble
+// of the first byte is zero in the case of an even number of nibbles and the first nibble
+// in the case of an odd number. All remaining nibbles (now an even number) fit properly
+// into the remaining bytes. Compact encoding is used for nodes stored on disk.
```
Comment (karalabe) [inline trie/encoding.go]: Ah sweet, I've never seen this part of the code in 2 years :))

File: trie/encoding_test.go [hunk 11]
```diff
@@ -17,113 +17,83 @@
...
+		{
+			key:    []byte{0x12, 0x34, 0x5},
+			hexIn:  []byte{1, 2, 3, 4, 0, 5, 16},
+			hexOut: []byte{1, 2, 3, 4, 0, 5, 16},
+		},
+		{
+			key:    []byte{0x12, 0x34, 0x56},
+			hexIn:  []byte{1, 2, 3, 4, 5, 6},
+			hexOut: []byte{1, 2, 3, 4, 5, 6, 16},
+		},
```
Comment (karalabe) [inline trie/encoding_test.go]: Could you add a test for encoding/decoding empty keys? The previous code had explicit checks whereas the new code does this implicitly by relying on the calculation returning an empty slice. There's even a slight API change, where the old code returned `nil` in `hexToKeybytes` whereas the new returns `[]byte{}`. I'd be happy to see an explicit test that verifies this weird case too.

File: trie/encoding_test.go [hunk 12]
```diff
@@ -17,113 +17,83 @@
...
+	tests := []struct{ hex, compact []byte }{
+		// even length, no terminator
+		{hex: []byte{1, 2, 3, 4, 5}, compact: []byte{0x11, 0x23, 0x45}},
+		// odd length, no terminator
+		{hex: []byte{0, 1, 2, 3, 4, 5}, compact: []byte{0x00, 0x01, 0x23, 0x45}},
+		// even length, terminator
+		{hex: []byte{0, 15, 1, 12, 11, 8, 16 /*term*/}, compact: []byte{0x20, 0x0f, 0x1c, 0xb8}},
+		// odd length, terminator
+		{hex: []byte{15, 1, 12, 11, 8, 16 /*term*/}, compact: []byte{0x3f, 0x1c, 0xb8}},
+	}
```
Comment (karalabe) [inline trie/encoding_test.go]: Could you add a test for encoding/decoding empty keys?

File: core/state/state_object.go [hunk 13]
```diff
@@ -280,7 +281,11 @@ func (c *stateObject) ReturnGas(gas *big.Int) {}
 
 func (self *stateObject) deepCopy(db *StateDB, onDirty func(addr common.Address)) *stateObject {
 	stateObject := newObject(db, self.address, self.data, onDirty)
-	stateObject.trie = self.trie
+	if self.trie != nil {
+		// A shallow copy makes the two tries independent.
+		cpy := *self.trie
+		stateObject.trie = &cpy
```
Comment (karalabe) [inline core/state/state_object.go]: Isn't `stateObject.trie` a `trie.SecureTrie`? Because that one is a much fancier construct that won't "deep copy" so easily https://github.com/ethereum/go-ethereum/blob/master/trie/secure_trie.go#L44

File: core/state/statedb.go [hunk 14]
```diff
@@ -296,6 +296,17 @@ func (self *StateDB) GetState(a common.Address, b common.Hash) common.Hash {
 	return common.Hash{}
 }
 
+// StorageTrie returns an iterator over storage trie of an account.
```
Comment (karalabe) [inline core/state/statedb.go]: Doesn't look like an iterator to me :P

File: eth/api.go [hunk 15]
```diff
@@ -526,59 +526,67 @@ func (api *PrivateDebugAPI) TraceTransaction(ctx context.Context, txHash common.
...
-		switch tracer := tracer.(type) {
-		case *vm.StructLogger:
-			return &ethapi.ExecutionResult{
-				Gas:         gas,
-				ReturnValue: fmt.Sprintf("%x", ret),
-				StructLogs:  ethapi.FormatLogs(tracer.StructLogs()),
-			}, nil
-		case *ethapi.JavascriptTracer:
-			return tracer.GetResult()
+			return nil, vm.Context{}, nil, fmt.Errorf("tx %x failed: %v", tx.Hash(), err)
```
Comment (karalabe) [inline eth/api.go]: Can this ever happen? If the tx is already in our chain, I think it's safe to say this won't ever fail. (Mentioning because you deleted a similar check at `tx.AsMessage`).

Comment (fjl) [inline eth/api.go]: The around AsMessage is for deriving the sender. I don't want to remove this one because many more things can go wrong (including DB errors, etc.).

File: core/state/state_object.go (hunk 13 above)
Comment (fjl) [inline core/state/state_object.go]: Shallow copy support for SecureTrie was added in [710435b51](https://github.com/ethereum/go-ethereum/commit/710435b51b97b4c688b70bda35ab9d1aa704a988#diff-6ed2f15666b9ea78bab176eaa5c69ba6R202), by yours truly.

File: trie/iterator.go (hunk 8 above)
Comment (fjl) [inline trie/iterator.go]: It has to return something. The idea was that the iterator wouldn't advance if there was an error fetching the next item.

File: core/state/state_object.go (hunk 13 above)
Comment (karalabe) [inline core/state/state_object.go]: Don't trust that guy.... he looks shady af ;)
